src/v/kafka/server/group_manager.cc

```diff
@@ -657,6 +658,62 @@ ss::future<> group_manager::recover_partition(
   model::term_id term,
   ss::lw_shared_ptr<attached_partition> p,
   group_recovery_consumer_state ctx) {
+    /*
+     * write the offset retention feature fence. if writing the fence fails
+     * recovery is not aborted. the effect of this is that availability is
+     * maintained a the expense of potentially having to manually delete
+     * committed offsets that are misclassified. however, since we just finished
+     * with a linearizable barrier and reading the partition, a failure here
+     * should be rare, and would be retried at the next leadership change.
+     */
+    if (!ctx.has_offset_retention_feature_fence) {
+        co_await _feature_table.local().await_feature(
+          features::feature::group_offset_retention, p->as);
+
+        auto batch = _feature_table.local().encode_version_fence();
+        auto reader = model::make_memory_record_batch_reader(std::move(batch));
+
+        try {
+            auto result = co_await p->partition->raft()->replicate(
+              term,
+              std::move(reader),
+              raft::replicate_options(raft::consistency_level::quorum_ack));
+            if (result) {
+                vlog(
+                  klog.info,
+                  "Wrote offset retention feature fence for partition {}",
+                  p->partition);
+            } else if (result.error() == raft::errc::shutting_down) {
+                vlog(
+                  klog.debug,
+                  "Cannot write offset retention version fence for partition "
+                  "{}: shutting down",
+                  p->partition);
+            } else if (result.error() == raft::errc::not_leader) {
+                vlog(
+                  klog.debug,
+                  "Cannot write offset retention version fence for partition "
+                  "{}: not leader",
+                  p->partition);
+            } else {
+                vlog(
+                  klog.warn,
+                  "Could not write offset retention feature fence for "
+                  "partition {}: {} {}",
+                  p->partition,
+                  result.error().message(),
+                  result.error());
+            }
+        } catch (...) {
+            vlog(
```

Comment:
1. Consumer groups will continue to work as expected even if writing the fence fails
2. Offset retention will even function correctly if writing the fence fails
3. When the fence fails to be written, the scenario of interest is that committed offsets that slip in before the fence does eventually get written will be misclassified as being from before the upgrade.

In the latest force push, the window of vulnerability for (3) is shrunk even more because we put the fence writing in a background retry loop. the backgrounding was required in any case because I forgot to take into account that I couldn't await_feature in this recovery path without making consumer groups unavailable during a partition upgrade.